File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        Res = [0] * len(nums)

        Prefix = 1 

        for i in range(len(nums)):
            Res[i] = Prefix
            Prefix = Prefix * nums[i]

        Postfix = 1 

        for i in range(len(nums)- 1 , -1 , -1 ):
            Res[i] = Res[i] * Postfix
            Postfix = Postfix * nums[i]
        
        return Res
    

        

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int res[] = new int[nums.length];

        int prefix = 1;

        for( int i = 0 ; i < nums.length ; i++)
        {
            res[i] = prefix;
            prefix = prefix * nums[i];
        }

        int postfix = 1;

        for(int i = nums.length - 1 ; i > -1 ; i--)
        {
            res[i] = res[i] * postfix;
            postfix = postfix * nums[i];
        }

        return res;
        
    }
}